### PR TITLE
Handle the situation where nprobe > nlist in IndexIVF

### PR DESCRIPTION
**Pull request by KinglittleQ**
## Description

Fix the bug mentioned in https://github.com/facebookresearch/faiss/issues/1010. When `nprobe` is greater than `nlist` in `IndexIVF`, the program will crash because the index will ask the quantizer to return more centroids than it owns.

## Changes:
1. Set `nprobe` as `nlist` if it is greater than `nlist` during searching.
2. Add one test to detect this bug.
3. Fix typo in `IndexPQ.cpp`.

### CODE REVIEW

File: faiss/IndexIVF.cpp [hunk 1]
```diff
@@ -277,10 +277,10 @@ void IndexIVF::set_direct_map_type (DirectMap::Type type)
 void IndexIVF::search (idx_t n, const float *x, idx_t k,
                          float *distances, idx_t *labels) const
 {
-
+    size_t nprobe = std::min(nlist, this->nprobe);
 
     // search function for a subset of queries
-    auto sub_search_func = [this, k]
+    auto sub_search_func = [this, k, nprobe]
```
Comment (mdouze) [inline faiss/IndexIVF.cpp:283]: this is useless

File: faiss/IndexIVF.cpp [hunk 2]
```diff
@@ -352,6 +352,8 @@ void IndexIVF::search_preassigned (idx_t n, const float *x, idx_t k,
                                    IndexIVFStats *ivf_stats) const
 {
     long nprobe = params ? params->nprobe : this->nprobe;
+    nprobe = std::min((long)nlist, nprobe);
```
Comment (mdouze) [inline faiss/IndexIVF.cpp:355]: looks good. While you are at it, could you replace occurrences of `long` with `idx_t` in this function ? We are trying to get rid of long, as its size is not portable.

**Review by mdouze — COMMENTED** [review]
Please address comments.

**Comment by KinglittleQ** [comment]
```
size_t nprobe = std::min(nlist, this->nprobe);
auto sub_search_func = [this, k, nprobe]
```
Here `nprobe` is a local variable instead of a member variable So it needs to be captured by the lambda function.

**Comment by KinglittleQ** [comment]
I found that calling `range_search` method would lead to a crash when `nprobe > nlist`. So I apply the same modification to `range_search` method.

New Changes:
1. The `range_search` method is also fixed.
2. Replace `long` with `idx_t` in `IndexIVF.cpp`.
3. Add a test to test `range_search`.

**Pull request closed without merging**